### PR TITLE
MONGOID-4528 Add more dirty methods

### PR DESCRIPTION
**Pull request by comandeo**
https://jira.mongodb.org/browse/MONGOID-4528

### CODE REVIEW

File: lib/mongoid/changeable.rb [hunk 1]
```diff
@@ -133,6 +135,46 @@ def setters
...
+    #
+    # @param [ String ] attr The name of the attribute.
+    # @param **kwargs The optional keyword arguments.
+    #
+    # @option **kwargs [ Object ] :from The object the attribute was changed from.
+    # @option **kwargs [ Object ] :to The object the attribute was changed to.
+    #
+    # @return [ true | false ] Whether the attribute change the next time we save.
+    def will_save_change_to_attribute?(attr, **kwargs)
+      attribute_changed?(attr, **kwargs)
```
Comment (comandeo) [inline lib/mongoid/changeable.rb:175]: I think it is safe to make `will_save_change_to_attribute?` basically an alias to `attribute_changed?`.
See https://github.com/rails/rails/issues/32694#issuecomment-383713284

File: docs/release-notes/mongoid-8.1.txt [hunk 2]
```diff
@@ -17,6 +17,48 @@ The complete list of releases is available `on GitHub
...
+
+  person = Person.create(name: 'John')
+  # "will_save_change_to_attribute?(:name): true"
+  # attribute_was(:name): John => New value
+  # attribute_before_last_save(:name): nil => Value before save
+  # saved_change_to_attribute(:name): [nil, "John"] => Both values
+  person.name = 'Jane'
+  person.save
+  # "will_save_change_to_attribute?(:name): true"
+  # attribute_was(:name): Jane => New value
```
Comment (Neilshweky) [inline docs/release-notes/mongoid-8.1.txt:53]: does attribute_was not look back before the last save? if not, can you mention that here?

File: lib/mongoid/changeable.rb [hunk 3]
```diff
@@ -133,6 +135,46 @@ def setters
...
+      attributes_before_last_save[attr]
+    end
+
+    # Returns the change to an attribute during the last save.
+    #
+    # @param [ Symbol | String ] attr The name of the attribute.
+    #
+    # @return [ Array<Object> | nil ] If the attribute was changed, returns
+    #   an array containing the original value and the saved value, otherwise nil.
+    def saved_change_to_attribute(attr)
```
Comment (Neilshweky) [inline lib/mongoid/changeable.rb:157]: can you also add the `saved_change_to_attribute?` method. I feel like that should be rather trivial to implement at this point.

File: lib/mongoid/changeable.rb (hunk 1 above)
Comment (Neilshweky) [reply lib/mongoid/changeable.rb:175]: yeah agreed.

**Review by Neilshweky — APPROVED** [review]

File: docs/release-notes/mongoid-8.1.txt [hunk 4]
```diff
@@ -17,6 +17,48 @@ The complete list of releases is available `on GitHub
...
+  # attribute_before_last_save(:name): John => Value before save
+  # saved_change_to_attribute(:name): ["John", "Jane"] => Both values
+
+For all of the new method there is also a shorter form created dynamically, e.g.
```
Comment (p-mongo) [inline docs/release-notes/mongoid-8.1.txt:57]:
```suggestion
For all of the new methods there are also shorter forms created dynamically, e.g.
```

File: spec/integration/callbacks_spec.rb [hunk 5]
```diff
@@ -447,4 +447,88 @@ class PreviouslyPersistedPerson
...
+
+      set_callback :save, :after do |doc|
+        [:name, :age].each do |attr|
+          saved_change_to_attribute_values_after[attr] += [saved_change_to_attribute(attr)]
+          attribute_before_last_save_values_after[attr] += [attribute_before_last_save(attr)]
+        end
+      end
+
+      def saved_change_to_attribute_values_before
+        @saved_change_to_attribute_values_before ||= Hash.new([])
```
Comment (p-mongo) [inline spec/integration/callbacks_spec.rb:474]: I think this copies the same array object into all new hash values which is generally not desirable. Unless this is specifically desired this should use the `new` with block syntax.

File: spec/integration/callbacks_spec.rb [hunk 6]
```diff
@@ -447,4 +447,88 @@ class PreviouslyPersistedPerson
...
+      def attribute_before_last_save_values_after
+        @attribute_before_last_save_values_after ||= Hash.new([])
+      end
+
+      def will_save_change_to_attribute_values_before
+        @will_save_change_to_attribute_values_before ||= Hash.new([])
+      end
+    end
+
+    it 'respoduces ActiveRecord::AttributeMethods::Dirty behavior' do
```
Comment (p-mongo) [inline spec/integration/callbacks_spec.rb:494]:
```suggestion
    it 'reproduces ActiveRecord::AttributeMethods::Dirty behavior' do
```

File: docs/release-notes/mongoid-8.1.txt (hunk 2 above)
Comment (comandeo) [reply docs/release-notes/mongoid-8.1.txt]: Done

File: lib/mongoid/changeable.rb (hunk 3 above)
Comment (comandeo) [reply lib/mongoid/changeable.rb:157]: Done

File: spec/integration/callbacks_spec.rb (hunk 5 above)
Comment (comandeo) [reply spec/integration/callbacks_spec.rb]: Done

**Review by p-mongo — APPROVED** [review]
I skimmed this but I don't see anything to object to.

**Pull request merged**